File: frontend/src/types/visaoGeral.ts
export interface VisaoGeralPeriodo {
  id: number
  data_inicio: string
  data_fim: string
  visitas: number
  visualizacoes_categoria: number
  visualizacoes_produto: number
  carrinhos_criados: number
  checkout_iniciado: number
  checkout_entrega: number
  checkout_pagamento: number
  pedidos_criados: number
  pedidos_pagos: number
  receita: string
  observacao: string
  vendas: number
  ticket_medio: string
  conversao_visitas_vendas: string
  conversao_visitas_carrinhos: string
  conversao_checkouts_vendas: string
  label_periodo: string
  criado_em: string
  atualizado_em: string
}

export interface VisaoGeralPeriodoInput {
  data_inicio: string
  data_fim: string
  visitas: number
  visualizacoes_categoria: number
  visualizacoes_produto: number
  carrinhos_criados: number
  checkout_iniciado: number
  checkout_entrega: number
  checkout_pagamento: number
  pedidos_criados: number
  pedidos_pagos: number
  receita: number
  observacao: string
}

File: backend/apps/finance/schemas/visao_geral.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from ninja import Schema


class VisaoGeralPeriodoIn(Schema):
    data_inicio: date
    data_fim: date
    visitas: int = 0
    visualizacoes_categoria: int = 0
    visualizacoes_produto: int = 0
    carrinhos_criados: int = 0
    checkout_iniciado: int = 0
    checkout_entrega: int = 0
    checkout_pagamento: int = 0
    pedidos_criados: int = 0
    pedidos_pagos: int = 0
    receita: Decimal = Decimal("0")
    observacao: str = ""


class VisaoGeralPeriodoOut(Schema):
    id: int
    data_inicio: date
    data_fim: date
    visitas: int
    visualizacoes_categoria: int
    visualizacoes_produto: int
    carrinhos_criados: int
    checkout_iniciado: int
    checkout_entrega: int
    checkout_pagamento: int
    pedidos_criados: int
    pedidos_pagos: int
    receita: Decimal
    observacao: str
    vendas: int
    ticket_medio: Decimal
    conversao_visitas_vendas: Decimal
    conversao_visitas_carrinhos: Decimal
    conversao_checkouts_vendas: Decimal
    label_periodo: str
    criado_em: datetime
    atualizado_em: datetime

    @staticmethod
    def resolve_label_periodo(obj) -> str:
        return (
            f"{obj.data_inicio.strftime('%d/%m/%y')} – "
            f"{obj.data_fim.strftime('%d/%m/%y')}"
        )


class VisaoGeralPeriodoPatch(Schema):
    data_inicio: Optional[date] = None
    data_fim: Optional[date] = None
    visitas: Optional[int] = None
    visualizacoes_categoria: Optional[int] = None
    visualizacoes_produto: Optional[int] = None
    carrinhos_criados: Optional[int] = None
    checkout_iniciado: Optional[int] = None
    checkout_entrega: Optional[int] = None
    checkout_pagamento: Optional[int] = None
    pedidos_criados: Optional[int] = None
    pedidos_pagos: Optional[int] = None
    receita: Optional[Decimal] = None
    observacao: Optional[str] = None
